pkg/typesystem/resolver.go

```diff
@@ -0,0 +1,102 @@
+package typesystem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/oklog/ulid/v2"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	modelSchemaVersionCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "model_schema_version_counter",
+			Help: "A counter counting the different authorization model schema versions that have been resolved",
+		},
+		[]string{"schema_version"},
+	)
+)
+
+// TypesystemResolverFunc is a function that implementations can implement to provide lookup and
+// resolution of a Typesystem.
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+
+// MemoizedTypesystemResolverFunc memoizes the provided TypesystemResolverFunc. It returns
+// the value from the first invocation of the underlying resolver function.
+//
+// The memoized resolver function is safe for concurrent use.
+func MemoizedTypesystemResolverFunc(fn TypesystemResolverFunc) TypesystemResolverFunc {
+	var cache sync.Map
```

Comment:
This may cause memory to grow overtime as it is not bounded nor has any timeout.  We may want to reconsider that will cause problem in a long running system